cloudconfig/util_test.go

```diff
@@ -0,0 +1,162 @@
+package cloudconfig
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/cloudbase/garm-provider-common/params"
+	"github.com/google/go-github/v55/github"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	newCloudCfg        = NewDefaultCloudInitConfig()
+	bootstrapParams = params.BootstrapInstance{
+		ExtraSpecs: []byte(`{"runner_install_template": "test"}`),
+		UserDataOptions: params.UserDataOptions{
+			DisableUpdatesOnBoot: newCloudCfg.PackageUpgrade,
+			ExtraPackages:        newCloudCfg.Packages,
+		},
+	}
+	tools = github.RunnerApplicationDownload{
+		Filename:          github.String("test-filename"),
+		DownloadURL:       github.String("https://example.com/test.zip"),
+		TempDownloadToken: github.String("test-token"),
+	}
+)
+
+func TestGetSpecs(t *testing.T) {
+	specs, err := GetSpecs(bootstrapParams)
+	require.NoError(t, err)
+	require.Equal(t, "\xb5\xeb-", string(specs.RunnerInstallTemplate))
```

Comment:
Hi @gabriel-samfira. Here I expected it to be equal to `test` instead of `\xb5\xeb-`. Do you have any idea why?

Reply:
`specs.RunnerInstallTemplate` is a `json.RawMessage{}`. You will have to:

```go
var data string
if err := json.Unmarshal(specs.RunnerInstallTemplate, &data); err != nil {
   // handle error
}

require.Equal(t, "test", data)
```

`specs.RunnerInstallTemplate` is not meant to be used by GARM, it's meant to be an opaque object that is meaningful only to a provider that implements it. Different providers may have different ideas in regards to what that install template may be.

Reply:
I updated the test with:
```go
specs, err := GetSpecs(bootstrapParams)
require.NoError(t, err)

var data string
if err := json.Unmarshal(specs.RunnerInstallTemplate, &data); err != nil {
	t.Fatalf("failed to unmarshal runner install template: %v", err)
}
require.Equal(t, "test", data)
```

and this is what I get after running it:
```go
failed to unmarshal runner install template: invalid character 'µ' looking for beginning of value
```

I'm not sure what's wrong here.

However, I noticed that by changing the `RunnerInstallTemplate` from `[]byte`:
https://github.com/cloudbase/garm-provider-common/blob/main/cloudconfig/util.go#L33 

to `json.RawMessage`, everything works as expected, and the JSON unmarshal doesn't fail anymore.

Reply:
try replacing `test` with `dGVzdAo=` (the base64 encoded version of `test`). In `CloudConfigSpec{}` the `RunnerInstallTemplate` field is a `[]byte{}`, which when marshaled is a base64 string.

Reply:
The idea is that a template is a file that may contain template tags. It's not a JSON. A file is typically represented as a byte array.

Reply:
It would be great to add an actual template there with some template tags and test that `ExtraContext` also does the right thing.